### DISCUSSION TITLE
Apply CS0282 to entire struct if applied to partial

### DISCUSSION THREAD

**Discussion by joelpryde** [dotnet/roslyn#53357] [Ideas] [answered]
We have been using source generators avidly to rework a lot of our more brittle codegen (and they are fantastic btw).  We use value types as much as possible in our codebase, and one issue we have run into is CS0282 warnings when we augment a partial struct with a source-generated one:

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8DEAHBCAc0gAIB3CBAOzWqNIBM0oJgM5SAGARgDYAnAFgAUAAEATD1GiA9LNIB5BGiJ0IGUrAQBXAMYxRYgMyk8VGGg1aYug6QCyATwAqTvJ1EBvUaT+kAMwwAewgYLWCwOAAxNDgMRgBuf1JRAF8ZEXlSAGVgnQQ9OABaIjhqRDC4Rhs7QxF8QhIICipaeiYWNg5uCQAOCSNTcwRLa219cOc3D1SRHxEUoNDwqEi4RRgAC0RY+KSU9KA==

Our options are as follows:
1. Add `/nowarn:0282` (the nuclear option, we are then hiding other - potentially valid - warnings in user code)
2. Add `[StructLayout(LayoutKind.Auto)]` (actually not an option for us as we rely on tooling down the line that needs to know the layout of our types from IL)
3. Place `#pragma warning disable 0282` on our generated struct

The third option seems by far the best.  Unfortunately Roslyn does not seem to apply this to the entire struct when it is place on one partial type of it.  Would it be possible to change the behavior to allow this?  Seems like the behavior that is most desirable from developers in any case and makes source generators more applicable for modifying partial struct types (attributes work this way already).

Let me know if this sounds amenable.  Happy to look into a PR from our side.

**Comment by RikkiGibson** [comment]
I think I'm sympathetic to wanting to enable source generators suppress warnings that they are "responsible" for. It feels like we would want to do a pass over the existing warnings that can apply to partial types and methods to decide if any others should get a similar treatment.

Perhaps we could [make `[SuppressMessage]` work on compiler warnings](https://github.com/dotnet/roslyn/issues/6471). I would expect adding such an attribute to suppress the warning at all declaration sites, since the attributes on a symbol are the union of all the attributes on all the partial declarations.

/cc @dotnet/roslyn-compiler for thoughts.

**Comment by PathogenDavid** [comment, marked as answer]
This seems like it'd be a good candidate for creating a [diagnostic suppressor](https://gist.github.com/mavasani/fcac17a9581b5c54cef8a689eeec954a). (That documentation and sample are from before the feature shipped, but I believe the API stayed the same.)

**Reply by RikkiGibson** [reply]
Awesome, I think my suggestion is moot in this case. I didn't realize we had this.

**Reply by joelpryde** [reply]
Interesting.  Would this work like an analyzer then that we would include on the csc command-line with /analyzer and it could then analyze a given instance of a diagnostic and choose to suppress it from being reported?

This is sort of the reverse of what I was looking for, since it will have to determine post-hoc if the warning was generated by our generator (I guess by looking at the symbol information of the generated type or looking for an attribute we place on it).  It could likely work for our purpose though as long as adding a specific suppressor does not add to compile time.

**Reply by PathogenDavid** [reply]
@RikkiGibson It's definitely a bit of a hidden feature! There's not even any samples for it in the roslyn-sdk repo.

@joelpryde Yup! `DiagnosticSuppressor` is actually a child of `DiagnosticAnalyzer`, so it literally is an analyzer.

You actually don't even need to create a separate assembly source source generators come from analyzer assemblies as far as Roslyn is concerned. IE: You can just ship them together in the same DLL and avoid making sure they're both referenced and compatible versions and all that.

If all you're doing is looking for structs marked with a specific attribute, I can't imagine this strategy would ever have a significant performance impact.
